### ISSUE TITLE
A node type was not returned.

### ISSUE THREAD

**Issue by zw-ruan** [twitter/torch-autograd#70]
Hi, 
I write a recursive autoencoder model, but get those message :

**A node type was not returned. This is either because a gradient was not defined, or the input is independent of the output.**

```lua
require 'torch'
grad = require 'autograd'


embeddingSize = 5

-- model parameters
-- encode parameters:
--      encodeW1, encodeW2, encodeB1
--
-- decode parameters
--      decodeW1, decodeW2, decodeB1, decodeB2
model_params = {
    W = {
        torch.randn(embeddingSize, embeddingSize), -- encodeW1
        torch.randn(embeddingSize, embeddingSize), -- encodeW2
        torch.randn(embeddingSize, embeddingSize), -- decodeW1
        torch.randn(embeddingSize, embeddingSize)  -- decodeW2
    },
    b = {
        torch.randn(embeddingSize, 1), -- encodeB1
        torch.randn(embeddingSize, 1), -- decodeB1
        torch.randn(embeddingSize, 1)  -- decodeB2
    }
}


--[[
    parameters:
        - input
            sentence embedding; the i-th column as w_i's enmbedding 
--]]
-- pass model_params to RAEModel
RAEModel = function(params, input)
    if input:size(2) == 1 then 
        return 0
    end

    local nodes = torch.Tensor(embeddingSize, input:size(2) * 2 - 1)
    local candidates_index = torch.range(1, input:size(2))
    candidates_index = candidates_index:long()
    nodes:indexCopy(2, candidates_index, input)
    local loss = 0.0
    -- run (sentence length - 1) turns
    for turn = 1, input:size(2) - 1 do
        local candidates = nodes:index(2, candidates_index)

        local C1 = candidates:narrow(2, 1, candidates:size(2) - 1) 
        local C2 = candidates:narrow(2, 2, candidates:size(2) - 1) 

        --     tanh(encodeW1 * C1 + encodeW2 * C2 + encodeB1)
        local P = torch.tanh(params.W[1] * C1 + params.W[2] * C2 +
                    torch.expand(params.b[1], C1:size(1), C1:size(2)))

        -- decode
        --      tanh(decodeW1 * P + decodeB1)
        --      tanh(decodeW2 * P + decodeB2)
        local C1Rec = torch.tanh(params.W[3] * P +
                        torch.expand(params.b[2], C1:size(1), C1:size(2)))

        local C2Rec = torch.tanh(params.W[4] * P +
                        torch.expand(params.b[3], C1:size(1), C1:size(2)))

        -- reconstruction error
        local recC1Error = C1Rec - C1
        local recC2Error = C2Rec - C2


        local recError = torch.sum(torch.pow(recC1Error, 2) +  
                                    torch.pow(recC2Error, 2), 1)

        -- get best pos 
        local mask = torch.eq(recError, torch.min(recError))[1]
        local index = torch.nonzero(mask)[1][1]
        loss = loss + recError[1][index]

        -- generate new candidates
        local p = P[{{}, {index}}]

        -- add p to nodes, at location input:size(2) + i
        nodes:indexCopy(2, torch.LongTensor{input:size(2) + turn}, p)

        -- Debug
        print('============' .. turn .. '===============')
        print('====> reconstruction error')
        print(recError[1])
        print('====> index: ' .. index)
        print('====> content in candidates_index: ')
        print(candidates_index)
        print('====> new generate content: ')
        print(nodes[{{}, input:size(2) + turn}])

        -- get the result, then return
        if turn == input:size(2) - 1 then
            print('====> nodes')
            print(nodes)
            return loss
        end


        -- generat new candidates index
        -- replace index and (index + 0) with (inputSize(2) + turn)
        -- if there is only one candidates, we get the result
        candidates_index[index] = input:size(2) + turn
        if index == 1 then
            torch.cat(candidates_index, candidates_index[{{index}}],
                                        candidates_index[{{index+2,
                                            candidates_index:size(1)}}])

        elseif index == candidates_index:size(1) - 1 then 
            torch.cat(candidates_index, candidates_index[{{1, 
                                                            index - 1}}],
                                        candidates_index[{{index}}])

        else
            torch.cat(candidates_index, candidates_index[{{1, 
                                                            index}}],
                            candidates_index[{{index+2, 
                                            candidates_index:size(1)}}])
        end

    end
end


input = torch.randn(5, 10)
dmodel = grad(RAEModel)
grad, loss = dmodel(model_params, input)

print('=========== Result ===============')
print('====> loss:')
print(loss)

```

**Comment by yongfei25** [comment]
Hi, your `RAEModel` function has to return the loss. In your case your loss is sitting in the `if` block, the loss is not returned if the block was not executed.

**Comment by zw-ruan** [comment]
I put a return statement before the end of function, still get the same error message.

**Comment by zw-ruan** [comment]
Hi, I change `RAEModel` function, get the same error message.
``` lua
RAEModel = function(params, input)
    local loss = 0.0
    if input:size(2) == 1 then 
        return loss
    end

    local nodes = torch.Tensor(embeddingSize, input:size(2) * 2 - 1)
    local candidates_index = torch.range(1, input:size(2))
    candidates_index = candidates_index:long()
    nodes:indexCopy(2, candidates_index, input)
    -- run (sentence length - 1) turns
    for turn = 1, input:size(2) - 1 do
        local candidates = nodes:index(2, candidates_index)

        local C1 = candidates:narrow(2, 1, candidates:size(2) - 1) 
        local C2 = candidates:narrow(2, 2, candidates:size(2) - 1) 

        --     tanh(encodeW1 * C1 + encodeW2 * C2 + encodeB1)
        local P = torch.tanh(params.W[1] * C1 + params.W[2] * C2 +
                    torch.expand(params.b[1], C1:size(1), C1:size(2)))

        -- decode
        --      tanh(decodeW1 * P + decodeB1)
        --      tanh(decodeW2 * P + decodeB2)
        local C1Rec = torch.tanh(params.W[3] * P +
                        torch.expand(params.b[2], C1:size(1), C1:size(2)))

        local C2Rec = torch.tanh(params.W[4] * P +
                        torch.expand(params.b[3], C1:size(1), C1:size(2)))

        -- reconstruction error
        local recC1Error = C1Rec - C1
        local recC2Error = C2Rec - C2

        -- apply 2-norm over column
        --local columnNorm = torch.norm(recC1Error, 2, 1)
        --recC1Error = torch.cdiv(recC1Error, 
        --                torch.expand(columnNorm, C1:size(1), C1:size(2)))
        --columnNorm = torch.norm(recC2Error, 2, 1)
        --recC2Error = torch.cdiv(recC2Error,
        --                torch.expand(columnNorm, C1:size(1), C1:size(2)))
        --

        -- local recError = torch.add(recC1Error, recC2Error)
        local recError = torch.sum(torch.pow(recC1Error, 2) +  
                                    torch.pow(recC2Error, 2), 1)

        -- get best pos 
        local mask = torch.eq(recError, torch.min(recError))[1]
        local index = torch.nonzero(mask)[1][1]
        loss = loss + recError[1][index]

        -- generate new candidates
        local p = P[{{}, {index}}]

        -- add p to nodes, at location input:size(2) + i
        nodes:indexCopy(2, torch.LongTensor{input:size(2) + turn}, p)

        -- Debug
        print('============' .. turn .. '===============')
        print('====> reconstruction error')
        print(recError[1])
        print('====> index: ' .. index)
        print('====> content in candidates_index: ')
        print(candidates_index)
        print('====> new generate content: ')
        print(nodes[{{}, input:size(2) + turn}])

        -- get the result, then return
        if turn == input:size(2) - 1 then
            print('====> nodes')
            print(nodes)
            break
        end


        -- generat new candidates index
        -- replace index and (index + 0) with (inputSize(2) + turn)
        -- if there is only one candidates, we get the result
        candidates_index[index] = input:size(2) + turn
        if index == 1 then
            torch.cat(candidates_index, candidates_index[{{index}}],
                                        candidates_index[{{index+2,
                                            candidates_index:size(1)}}])

        elseif index == candidates_index:size(1) - 1 then 
            torch.cat(candidates_index, candidates_index[{{1, 
                                                            index - 1}}],
                                        candidates_index[{{index}}])

        else
            torch.cat(candidates_index, candidates_index[{{1, 
                                                            index}}],
                            candidates_index[{{index+2, 
                                            candidates_index:size(1)}}])
        end

    end

    return loss
end
```

**Comment by yongfei25** [comment]
I guess if you init `loss` inside the function, then `loss` is not a node type. That's why it is complaining.
```lua
local loss = 0.0 -- Not a node type
...
return loss  
```

If you want to init `loss` to 0, you can pass it into the grad function as one of the arguments.
``` lua
RAEModel = function(params, input, loss)
  ...
  return loss
end

df = grad(RAEModel)
df(params, input, 0)
```

**Comment by zw-ruan** [comment]
@Yong Fei
Thanks. But it is still complaining.

2015-12-24 0:35 GMT+08:00 Yong Fei <notifications@github.com>:

> I guess if you init loss inside the function, then loss is not a node
> type. That's why it is complaining.
>
> local loss = 0.0 -- Not a node type...return loss
>
> If you want to init loss to 0, you can pass it into the grad function as
> one of the arguments.
>
> RAEModel = function(params, input, loss)
>   ...
>   return lossend
>
> df = grad(RAEModel)df(params, input, 0)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/twitter/torch-autograd/issues/70#issuecomment-166938259>
> .
>



-- 
阮志伟
厦门大学信息科学与技术学院
智能科学与技术系 15级硕士生
电话：18650127769

**Comment by zw-ruan** [comment]
I change all object-oriented function calls to functions in autograd package support.lua. But it still complains.
``` lua
RAEModel = function(params, input)
    -- if input:size(2) == 1 then 
    local loss = 0.0 
    if torch.size(input, 2) == 1 then 
        return input
        -- return loss
    end

    local inputSize = torch.size(input, 2)
    -- local nodes = torch.Tensor(embeddingSize, input:size(2) * 2 - 1)
    local nodes = torch.Tensor(embeddingSize, inputSize * 2 - 1)
    local candidates_index = torch.range(1, inputSize)
    candidates_index = torch.long(candidates_index)
    -- nodes:indexCopy(2, candidates_index, input)
    -- torch.narrow(nodes, 2, 1, torch.size(input, 2)).copy(input)
    nodes[{{}, {1, inputSize}}] = input
    
    -- run (sentence length - 1) turns
    for turn = 1, torch.size(input, 2) - 1 do
        -- return new tensor, not memory friendly
        local candidates = torch.index(nodes, 2, candidates_index)
        
        local candidates_size = torch.size(candidates, 2)
        local C1 = torch.narrow(candidates, 2, 1, candidates_size - 1) 
        local C2 = torch.narrow(candidates, 2, 2, candidates_size - 1)

        --     tanh(encodeW1 * C1 + encodeW2 * C2 + encodeB1)
        local P = torch.tanh(params.W[1] * C1 + params.W[2] * C2 +
                    torch.expand(params.b[1], torch.size(C1,1), 
                                                torch.size(C1,2)))

        -- decode
        --      tanh(decodeW1 * P + decodeB1)
        --      tanh(decodeW2 * P + decodeB2)
        local C1Rec = torch.tanh(params.W[3] * P +
                        torch.expand(params.b[2], torch.size(C1, 1), 
                                                torch.size(C1, 2)))

        local C2Rec = torch.tanh(params.W[4] * P +
                        torch.expand(params.b[3], torch.size(C1, 1), 
                                                torch.size(C1, 2)))

        -- reconstruction error
        local recC1Error = C1Rec - C1
        local recC2Error = C2Rec - C2

        -- apply 2-norm over column
        --local columnNorm = torch.norm(recC1Error, 2, 1)
        --recC1Error = torch.cdiv(recC1Error, 
        --                torch.expand(columnNorm, C1:size(1), C1:size(2)))
        --columnNorm = torch.norm(recC2Error, 2, 1)
        --recC2Error = torch.cdiv(recC2Error,
        --                torch.expand(columnNorm, C1:size(1), C1:size(2)))
        --

        -- local recError = torch.add(recC1Error, recC2Error)
        local recError = torch.sum(torch.pow(recC1Error, 2) +  
                                    torch.pow(recC2Error, 2), 1)

        -- get best pos 
        local mask = torch.eq(recError, torch.min(recError))[1]
        local index = torch.nonzero(mask)[1][1]
        loss = loss + recError[1][index]

        -- generate new candidates
        local p = P[{{}, {index}}]
        -- add p to nodes, at location input:size(2) + i
        nodes[{{}, {input:size(2) + turn}}] = p

        -- Debug
        print('============' .. turn .. '===============')
        print('====> reconstruction error')
        print(recError[1])
        print('====> index: ' .. index)
        print('====> content in candidates_index: ')
        print(candidates_index)
        print('====> new generate content: ')
        print(nodes[{{}, input:size(2) + turn}])

        -- get the result, then return
        if turn == input:size(2) - 1 then
            print('====> nodes')
            print(nodes)
            break
        end


        -- generat new candidates index
        -- replace index and (index + 0) with (inputSize(2) + turn)
        -- if there is only one candidates, we get the result
        candidates_index[index] = inputSize + turn
        if index == 1 then
            torch.cat(candidates_index, candidates_index[{{index}}],
                                        candidates_index[{{index+2,
                                        torch.size(candidates_index, 1)}}])

        elseif index == candidates_index:size(1) - 1 then 
            torch.cat(candidates_index, candidates_index[{{1, 
                                                            index - 1}}],
                                        candidates_index[{{index}}])

        else
            torch.cat(candidates_index, candidates_index[{{1, 
                                                            index}}],
                            candidates_index[{{index+2, 
                                        torch.size(candidates_index, 1)}}])
        end

    end

    return loss
end
```

**Comment by alexbw** [comment]
You cannot overwrite a variables value in autograd. Do you redefine or
update loss at any point?

**Comment by alexbw** [comment]
So anything of the form loss = loss + a is invalid

**Comment by zw-ruan** [comment]
@alexbw 
What if I accumulate loss with the other variable and assign to loss finally?

**Comment by zw-ruan** [comment]
I get it. So I cannot overwrite any variable in autograd.

**Comment by alexbw** [comment]
Correct. We currently don't detect explicitly when you overwrite variables,
which might not be possible without analyzing the source code of a
differentiated function. We do not have that capability currently.

**Comment by zw-ruan** [comment]
Does autograd support recursive function, i.e.
``` lua
function model(params, input)
    if(condition)
         return something
    return something + model(params, newInput)
end
```
